lib/stages/VisSharedMemWriter.cpp

```diff
@@ -0,0 +1,379 @@
+
+    // Acquire semaphore until shared memory is created
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {}", _sem_name);
+        return;
+    }
+
+    // Set up the structure of the ring buffer shared memory
+    // Get one frame for reference
+    wait_for_full_frame(in_buf, unique_name.c_str(), frame_id);
+
+    auto frame = visFrameView(in_buf, frame_id);
+
+    // Build the frequency index
+    std::map<uint32_t, uint32_t> freq_id_map;
+    auto& dm = datasetManager::instance();
+
+    const freqState* freq_state = dm.dataset_state<freqState>(frame.dataset_id);
+
+    uint ind = 0;
+    for (auto& f : freq_state->get_freqs())
+        freq_id_map[f.first] = ind++;
+
+
+    // Figure out the ring buffer structure
+    nfreq = freq_state->get_freqs().size();
+
+    // Set the alignment (in kB)
+    size_t alignment = 4096; // Align on page boundaries
+
+    // Calculate the ring buffer structure
+
+    data_size = frame.data_size;
+    metadata_size = sizeof(visMetadata);
+    // Alligns the frame along page size
+    frame_size = _member_alignment(data_size + metadata_size + valid_size, alignment);
+
+    // memory_size should be _ntime * nfreq * file_frame_size (data + metadata)
+    buf_addr = assign_memory(_fname_buf, (structured_data_size * structured_data_num) + (_ntime * nfreq * access_record_size) + (_ntime * nfreq * frame_size));
+
+    // The elements contained in the structured data and access record are each 64 bytes
+    structured_data_addr = (uint64_t*) buf_addr;
+    access_record_addr = (int64_t*) (structured_data_addr + structured_data_num);
+    buf_addr += (structured_data_size * structured_data_num) + (_ntime * nfreq * access_record_size);
+
+    // Record structure of data
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
+    memcpy(structured_data_addr + 1, &_ntime, sizeof(_ntime));
+    memcpy(structured_data_addr + 2, &nfreq, sizeof(nfreq));
+    memcpy(structured_data_addr + 3, &frame_size, sizeof(frame_size));
+    memcpy(structured_data_addr + 4, &metadata_size, sizeof(metadata_size));
+    memcpy(structured_data_addr + 5, &data_size, sizeof(data_size));
+
+    // initially set the address records with -1
+    // std::fill_n(array, 100, -1);
+    int64_t invalid_vector [_ntime * nfreq];
+    std::fill_n(invalid_vector, _ntime*nfreq, invalid);
+    memcpy(access_record_addr, invalid_vector, _ntime * nfreq * sizeof(invalid_vector[0]));
+
+    INFO("Created the shared memory segments\n");
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+
+    // gets called once when kotekan is running
+    while (!stop_thread) {
+
+
+        // Wait for the buffer to be filled with data
+        if (wait_for_full_frame(in_buf, unique_name.c_str(), frame_id) == nullptr) {
+            break;
+        }
+
+        // Get a view of the current frame
+        auto frame = visFrameView(in_buf, frame_id);
+
```

Comment:
I'd like there to be some code in here which creates a fingerprint of a set of specified dataset states (inputs, prods, stack, freq, metadata + specified others) and asserts that it hasn't changed. That's done in a few places (e.g. visWriter), so you can crib from there, but the general pattern is.

- First frame:
  * calculate the fingerprint with respect to the interesting states and save it
  * insert the dataset_id of the frame into a set.
- All other frames:
  * check if the dataset_id is in the set:
    * if it is, do nothing and write the data
    * if it is not:
      * calculate the fingerprint of the new dataset, and check that it matches the saved fingerprint
      * if it does, add the dataset_id to the set, and write the data.
      * if it does not, throw an error as the buffer is now receiving data in a way you guaranteed it would.